File: lib/permissions.ts
import { Role } from "@prisma/client";

// ─── Permission Definitions ──────────────────────────────

export interface Permission {
  // Portal access
  canAccessOpsPortal: boolean;
  canAccessSchoolPortal: boolean;

  // School management
  canManageSchools: boolean;
  canViewAllSchools: boolean;
  canEditFeatureFlags: boolean;

  // Camera operations
  canViewCameras: boolean;
  canManageCameras: boolean;
  canViewLiveFeeds: boolean;

  // Alert operations
  canViewAlerts: boolean;
  canResolveAlerts: boolean;
  canViewAllAlerts: boolean; // cross-school

  // User management
  canManageUsers: boolean;
  canManageAllUsers: boolean; // cross-school

  // Dashboard
  canViewDashboard: boolean;
  canViewOpsDashboard: boolean;
}

// ─── Role → Permission Map ───────────────────────────────

const ROLE_PERMISSIONS: Record<Role, Permission> = {
  SUPER_ADMIN: {
    canAccessOpsPortal: true,
    canAccessSchoolPortal: true,
    canManageSchools: true,
    canViewAllSchools: true,
    canEditFeatureFlags: true,
    canViewCameras: true,
    canManageCameras: true,
    canViewLiveFeeds: true,
    canViewAlerts: true,
    canResolveAlerts: true,
    canViewAllAlerts: true,
    canManageUsers: true,
    canManageAllUsers: true,
    canViewDashboard: true,
    canViewOpsDashboard: true,
  },
  OPS_VIEWER: {
    canAccessOpsPortal: true,
    canAccessSchoolPortal: false,
    canManageSchools: false,
    canViewAllSchools: true,
    canEditFeatureFlags: false,
    canViewCameras: true,
    canManageCameras: false,
    canViewLiveFeeds: false, // OPS_VIEWER cannot view live feeds
    canViewAlerts: true,
    canResolveAlerts: false,
    canViewAllAlerts: true,
    canManageUsers: false,
    canManageAllUsers: false,
    canViewDashboard: false,
    canViewOpsDashboard: true,
  },
  SCHOOL_ADMIN: {
    canAccessOpsPortal: false,
    canAccessSchoolPortal: true,
    canManageSchools: false,
    canViewAllSchools: false,
    canEditFeatureFlags: false,
    canViewCameras: true,
    canManageCameras: false, // read-only camera management
    canViewLiveFeeds: true,
    canViewAlerts: true,
    canResolveAlerts: true,
    canViewAllAlerts: false,
    canManageUsers: true, // within their school only
    canManageAllUsers: false,
    canViewDashboard: true,
    canViewOpsDashboard: false,
  },
  SCHOOL_VIEWER: {
    canAccessOpsPortal: false,
    canAccessSchoolPortal: true,
    canManageSchools: false,
    canViewAllSchools: false,
    canEditFeatureFlags: false,
    canViewCameras: true,
    canManageCameras: false,
    canViewLiveFeeds: true,
    canViewAlerts: true,
    canResolveAlerts: false,
    canViewAllAlerts: false,
    canManageUsers: false,
    canManageAllUsers: false,
    canViewDashboard: true,
    canViewOpsDashboard: false,
  },
};

// ─── Helper Functions ────────────────────────────────────

/**
 * Get the full permission set for a role
 */
export function getPermissions(role: Role): Permission {
  return ROLE_PERMISSIONS[role];
}

/**
 * Check if a role has a specific permission
 */
export function hasPermission(
  role: Role,
  permission: keyof Permission
): boolean {
  return ROLE_PERMISSIONS[role][permission];
}

/**
 * Check if a role is an ops-level role (internal team)
 */
export function isOpsRole(role: Role): boolean {
  return role === "SUPER_ADMIN" || role === "OPS_VIEWER";
}

/**
 * Check if a role is a school-level role
 */
export function isSchoolRole(role: Role): boolean {
  return role === "SCHOOL_ADMIN" || role === "SCHOOL_VIEWER";
}

/**
 * Check if a role can access a specific school's data
 */
export function canAccessSchoolData(
  role: Role,
  userSchoolId: string | null,
  targetSchoolId: string
): boolean {
  // Ops roles can access any school
  if (isOpsRole(role)) return true;
  // School roles can only access their own school
  return userSchoolId === targetSchoolId;
}

/**
 * Check if a role can manage users with a target role.
 * SCHOOL_ADMIN can only create/edit SCHOOL_ADMIN and SCHOOL_VIEWER.
 * SUPER_ADMIN can create/edit any role.
 */
export function canManageRole(
  managerRole: Role,
  targetRole: Role
): boolean {
  if (managerRole === "SUPER_ADMIN") return true;
  if (managerRole === "SCHOOL_ADMIN") {
    return targetRole === "SCHOOL_ADMIN" || targetRole === "SCHOOL_VIEWER";
  }
  return false;
}

/**
 * Get the roles that a manager role can assign
 */
export function assignableRoles(managerRole: Role): Role[] {
  if (managerRole === "SUPER_ADMIN") {
    return ["SUPER_ADMIN", "OPS_VIEWER", "SCHOOL_ADMIN", "SCHOOL_VIEWER"];
  }
  if (managerRole === "SCHOOL_ADMIN") {
    return ["SCHOOL_ADMIN", "SCHOOL_VIEWER"];
  }
  return [];
}

/**
 * Get the default redirect path after login based on role
 */
export function getLoginRedirect(role: Role): string {
  if (isOpsRole(role)) return "/ops/dashboard";
  return "/dashboard";
}

// ─── API Route Helpers ───────────────────────────────────

/**
 * Require a specific permission. Throws if not met.
 * Use in API routes: requirePermission(session.user.role, 'canManageUsers')
 */
export function requirePermission(
  role: Role,
  permission: keyof Permission
): void {
  if (!hasPermission(role, permission)) {
    throw new Error(
      `Forbidden: role ${role} does not have permission ${permission}`
    );
  }
}

/**
 * Require that the user can access a specific school.
 * Throws if the user is school-scoped and the IDs don't match.
 */
export function requireSchoolAccess(
  role: Role,
  userSchoolId: string | null,
  targetSchoolId: string
): void {
  if (!canAccessSchoolData(role, userSchoolId, targetSchoolId)) {
    throw new Error(
      `Forbidden: user cannot access school ${targetSchoolId}`
    );
  }
}

File: dashboard-go/internal/permissions/permissions.go
package permissions

// Permission defines the full set of boolean capabilities for a role.
type Permission struct {
	CanAccessOpsPortal    bool
	CanAccessSchoolPortal bool
	CanManageSchools      bool
	CanViewAllSchools     bool
	CanEditFeatureFlags   bool
	CanViewCameras        bool
	CanManageCameras      bool
	CanViewLiveFeeds      bool
	CanViewAlerts         bool
	CanResolveAlerts      bool
	CanViewAllAlerts      bool
	CanManageUsers        bool
	CanManageAllUsers     bool
	CanViewDashboard      bool
	CanViewOpsDashboard   bool
}

// RolePermissions maps each role name to its complete permission set.
// Values match the TypeScript ROLE_PERMISSIONS exactly.
var RolePermissions = map[string]Permission{
	"SUPER_ADMIN": {
		CanAccessOpsPortal:    true,
		CanAccessSchoolPortal: true,
		CanManageSchools:      true,
		CanViewAllSchools:     true,
		CanEditFeatureFlags:   true,
		CanViewCameras:        true,
		CanManageCameras:      true,
		CanViewLiveFeeds:      true,
		CanViewAlerts:         true,
		CanResolveAlerts:      true,
		CanViewAllAlerts:      true,
		CanManageUsers:        true,
		CanManageAllUsers:     true,
		CanViewDashboard:      true,
		CanViewOpsDashboard:   true,
	},
	"OPS_VIEWER": {
		CanAccessOpsPortal:    true,
		CanAccessSchoolPortal: false,
		CanManageSchools:      false,
		CanViewAllSchools:     true,
		CanEditFeatureFlags:   false,
		CanViewCameras:        true,
		CanManageCameras:      false,
		CanViewLiveFeeds:      false,
		CanViewAlerts:         true,
		CanResolveAlerts:      false,
		CanViewAllAlerts:      true,
		CanManageUsers:        false,
		CanManageAllUsers:     false,
		CanViewDashboard:      false,
		CanViewOpsDashboard:   true,
	},
	"SCHOOL_ADMIN": {
		CanAccessOpsPortal:    false,
		CanAccessSchoolPortal: true,
		CanManageSchools:      false,
		CanViewAllSchools:     false,
		CanEditFeatureFlags:   false,
		CanViewCameras:        true,
		CanManageCameras:      false,
		CanViewLiveFeeds:      true,
		CanViewAlerts:         true,
		CanResolveAlerts:      true,
		CanViewAllAlerts:      false,
		CanManageUsers:        true,
		CanManageAllUsers:     false,
		CanViewDashboard:      true,
		CanViewOpsDashboard:   false,
	},
	"SCHOOL_VIEWER": {
		CanAccessOpsPortal:    false,
		CanAccessSchoolPortal: true,
		CanManageSchools:      false,
		CanViewAllSchools:     false,
		CanEditFeatureFlags:   false,
		CanViewCameras:        true,
		CanManageCameras:      false,
		CanViewLiveFeeds:      true,
		CanViewAlerts:         true,
		CanResolveAlerts:      false,
		CanViewAllAlerts:      false,
		CanManageUsers:        false,
		CanManageAllUsers:     false,
		CanViewDashboard:      true,
		CanViewOpsDashboard:   false,
	},
}

// permissionField maps a permission name string to a getter function on Permission.
var permissionField = map[string]func(Permission) bool{
	"canAccessOpsPortal":    func(p Permission) bool { return p.CanAccessOpsPortal },
	"canAccessSchoolPortal": func(p Permission) bool { return p.CanAccessSchoolPortal },
	"canManageSchools":      func(p Permission) bool { return p.CanManageSchools },
	"canViewAllSchools":     func(p Permission) bool { return p.CanViewAllSchools },
	"canEditFeatureFlags":   func(p Permission) bool { return p.CanEditFeatureFlags },
	"canViewCameras":        func(p Permission) bool { return p.CanViewCameras },
	"canManageCameras":      func(p Permission) bool { return p.CanManageCameras },
	"canViewLiveFeeds":      func(p Permission) bool { return p.CanViewLiveFeeds },
	"canViewAlerts":         func(p Permission) bool { return p.CanViewAlerts },
	"canResolveAlerts":      func(p Permission) bool { return p.CanResolveAlerts },
	"canViewAllAlerts":      func(p Permission) bool { return p.CanViewAllAlerts },
	"canManageUsers":        func(p Permission) bool { return p.CanManageUsers },
	"canManageAllUsers":     func(p Permission) bool { return p.CanManageAllUsers },
	"canViewDashboard":      func(p Permission) bool { return p.CanViewDashboard },
	"canViewOpsDashboard":   func(p Permission) bool { return p.CanViewOpsDashboard },
}

// HasPermission checks whether the given role has the named permission.
// Returns false for unknown roles or permission names.
func HasPermission(role, permission string) bool {
	perms, ok := RolePermissions[role]
	if !ok {
		return false
	}
	getter, ok := permissionField[permission]
	if !ok {
		return false
	}
	return getter(perms)
}

// IsOpsRole returns true if the role is an internal operations role.
func IsOpsRole(role string) bool {
	return role == "SUPER_ADMIN" || role == "OPS_VIEWER"
}

// IsSchoolRole returns true if the role is a school-scoped role.
func IsSchoolRole(role string) bool {
	return role == "SCHOOL_ADMIN" || role == "SCHOOL_VIEWER"
}

// CanAccessSchoolData checks whether a user with the given role and school ID
// is permitted to access data for the target school.
// Ops roles can access any school; school roles can only access their own.
func CanAccessSchoolData(role, userSchoolID, targetSchoolID string) bool {
	if IsOpsRole(role) {
		return true
	}
	return userSchoolID == targetSchoolID
}

// CanManageRole checks whether a manager with the given role can create or edit
// users with the target role.
// SUPER_ADMIN can manage any role. SCHOOL_ADMIN can manage SCHOOL_ADMIN and
// SCHOOL_VIEWER. All other roles cannot manage anyone.
func CanManageRole(managerRole, targetRole string) bool {
	if managerRole == "SUPER_ADMIN" {
		return true
	}
	if managerRole == "SCHOOL_ADMIN" {
		return targetRole == "SCHOOL_ADMIN" || targetRole == "SCHOOL_VIEWER"
	}
	return false
}

// AssignableRoles returns the list of roles that a manager with the given role
// is allowed to assign to other users.
func AssignableRoles(managerRole string) []string {
	switch managerRole {
	case "SUPER_ADMIN":
		return []string{"SUPER_ADMIN", "OPS_VIEWER", "SCHOOL_ADMIN", "SCHOOL_VIEWER"}
	case "SCHOOL_ADMIN":
		return []string{"SCHOOL_ADMIN", "SCHOOL_VIEWER"}
	default:
		return []string{}
	}
}

// GetLoginRedirect returns the default path to redirect to after login
// based on the user's role.
func GetLoginRedirect(role string) string {
	if IsOpsRole(role) {
		return "/ops/dashboard"
	}
	return "/dashboard"
}
